src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -361,8 +370,12 @@ double SolverInterfaceImpl::advance(
   Event                    e("advance", precice::syncMode);
   utils::ScopedEventPrefix sep("advance/");
 
-  PRECICE_CHECK(_couplingScheme->isInitialized(), "initialize() has to be called before advance()");
-  PRECICE_CHECK(isCouplingOngoing(), "advance() cannot be called when isCouplingOngoing() returns false");
+  PRECICE_CHECK(_state != State::Constructed, "initialize() has to be called before advance().");
+  PRECICE_CHECK(_state != State::Finalized, "advance() cannot be called after finalize().")
+  PRECICE_ASSERT(_couplingScheme->isInitialized());
+  PRECICE_CHECK(isCouplingOngoing(), "advance() cannot be called when isCouplingOngoing() returns false.");
+  PRECICE_CHECK((not _couplingScheme->receivesInitializedData() && not _couplingScheme->sendsInitializedData()) || (_state == State::InitializedData),
+                "initializeData() needs to be called before advance if data has to be initialized.");
```

Comment:
Can you add the isCouplingOngoing() and reword the coupling data parts in the preconditions part of the documentation? (See SolverInterface.hpp)

Reply:
Does https://github.com/precice/precice/commit/eab65f3d394dfb05c1b44e0b8901e92b4e997a12 look ok?